### ISSUE TITLE
guest room settings screen: sort order of devices seems arbitrary

### ISSUE THREAD

**Issue by chrismarquardt** [steveseguin/obsninja#412]
sort order of devices in guest settings screen seems arbitrary, probably better to have them be alphabetical

<img width="364" alt="devices_sortorder" src="https://user-images.githubusercontent.com/2527608/96372992-f4b9fd00-1169-11eb-8611-a49986f99263.png">

**Comment by steveseguin** [comment]
![image](https://user-images.githubusercontent.com/2575698/96380207-8d964b80-115d-11eb-88ed-e737e3f2afe4.png)
![image](https://user-images.githubusercontent.com/2575698/96380347-97b84a00-115d-11eb-806a-7def9d3d7142.png)

NDI will always be Last in the list, unless manually selected via something like &device=ndi

**Comment by steveseguin** [comment]
I'm thinking I should still leave the first device alone -- as if its the default, shoudln't I leave it that way?

**Comment by chrismarquardt** [comment]
The first device is not the default on my system. That should be the built-in FaceTime camera, but it is the Elgato. Arbitrary. Maybe Mac-specific issue?

**Comment by chrismarquardt** [comment]
Here's beta on my Mac
<img width="494" alt="Screenshot 2020-10-18 at 22 48 09" src="https://user-images.githubusercontent.com/2527608/96385406-02847800-1194-11eb-915d-093889c97b7a.png">

**Comment by steveseguin** [comment]
The audio device is "Default"

**Comment by chrismarquardt** [comment]
You can set default devices for audio in and out on a Mac, but you can't set default video devices. At least there's no interface for it.

**Comment by steveseguin** [comment]
should I leave the audio device that is default on top?  Should I select the default device by default?  should I alpha-order the device device, too, but still select it?

**Comment by chrismarquardt** [comment]
Most of my guests will have one or two mics (one as default) and one camera, so selecting the audio default is fine and selecting the first camera in the list is probably fine too. For me personally it's a bit annoying, because for quick tests, I use the Facetime cam.

Maybe this might work on Mac:

1. alpha-sort all three lists (audio in/out, video in)
2. select audio devices based on what's set as default in the system
3. keep the last selected camera in a cookie and select that on return

Especially point 3 would add some consistency between visits for Mac users

**Comment by steveseguin** [comment]
This is all far more complex than it seems, for numerous reasons. haha.

I've commented-out the changes I made, as I think this needs more time and testing, which isn't appropriate for the state of beta currently.

Just for example, when I request camera permissions, I need to use "auto" as the source for audio and video, since I don't have device IDs at this point.  

This default device picked is the first device in the device list that I've yet to gain access to.  Now that I have that stream, I can get a list of all devices, but having to change it to another device at this point just means even more time and complication. 

I get more stability if I just go with the defaults (usually assigned by Chrome/Windows), and avoid changing the device unless needed by the user. Changing between video sources in an few milliseconds can cause issues as well, where some devices jam up, take time to change tracks, or crash some devices. 

anyways,  I think the cookie idea could work, or make sense to implement, but I've so far avoided cookies.

**Comment by chrismarquardt** [comment]
yeah, always more complex under the hood. and i didn't want this in beta/production, this is clearly dev stuff and also not a showstopper, just a usability thing

**Comment by steveseguin** [comment]
Closing old issue automatically by bot; device lists are now sorted by label.
